src/host/allocations_cache.jl

```diff
@@ -0,0 +1,251 @@
+    named_cache = get(dev_cache, name, nothing)
+    named_cache ≡ nothing && return
+
+    Base.@lock named_cache.lock begin
+        for (_, pool) in named_cache.free
+            map(unsafe_free!, pool)
+        end
+        # TODO error when trying to invalidate busy cache?
+        for (_, pool) in named_cache.busy
+            map(unsafe_free!, pool)
+        end
+        empty!(named_cache.busy)
+        empty!(named_cache.free)
+    end
+    return
+end
+
+function free_busy!(kab::Backend, name::Symbol)
+    pdcache = cache_allocator(kab)
+    free_busy!(named_cache_allocator!(pdcache, device(kab), name); pdcache.free_immediately)
+end
+
+"""
+    @cache_scope backend name expr
+
+Evaluate expression `expr` using `name`d caching allocator
+for the given KernelAbstractions `backend`.
+
+When gpu allocation is requested during execution of `expr`,
+allocator will try to use its "free" cache instead of doing an actual allocation.
+If no "free" allocation exists, an actual allocation is performed.
+Before returning allocation to the user, it is marked as busy and
+will not be used by allocation in the scope defined by `@cache_scope`.
+
+**After** the execution of `expr` all "busy" allocations are marked as "free"
+thus they can be re-used next time the program enters this scope.
+
+This is useful to apply in a repeating block of code to avoid relying on
+GC to free gpu memory in time.
+
+`name` is a `Symbol` that defines which allocator to use
+(`:none` is reserved and means no allocator).
+
+# Example
+
+In the following example, each iteration of the for-loop requires `2 GiB`
+of gpu memory.
+Without caching allocator GC wouldn't be able to free arrays in time
+resulting in higher memory usage.
+With caching allocator, memory usage stays at exactly `2 GiB`.
+
+See [`@no_cache_scope`](@ref), [`invalidate_cache_allocator!`](@ref).
+```julia
+kab = CUDABackend()
+n = 1024^3
+for i in 1:1000
+    @cache_scope kab :loop begin
+        sin.(CUDA.rand(Float32, n))
+    end
+end
+invalidate_cache_allocator!(kab, :loop)
+```
+"""
+macro cache_scope(backend, name, expr)
+    quote
+        res = @with $(esc(CacheAllocatorName)) => $(esc(name)) $(esc(expr))
+        free_busy!($(esc(backend)), $(esc(name)))
+        res
+    end
+end
+
+"""
+    @no_cache_scope expr
+
+Evaluate expression `expr` without using caching allocator.
+This is useful to call from within `@cache_scope` to avoid caching arrays.
+"""
+macro no_cache_scope(expr)
+    quote
+        @with $(esc(CacheAllocatorName)) => :none $(esc(expr))
+    end
+end
+
+# Interface API.
+
+"""
+    cache_allocator(::Backend)
+
+Given KernelAbstractions `backend`, return corresponding `PerDeviceCacheAllocator` for it.
+Each GPU backend must implement this.
+"""
+cache_allocator(::Backend) = error("Not implemented.")
+
+"""
+    device(::Backend)
+
+Given KernelAbstractions `backend`, return current device.
+Each GPU backend must implement this.
+"""
+device(::Backend) = error("Not implemented.")
```

Comment:
This seems like something that should go in KernelAbstractions.jl itself?

Alternatively, when keying on the array type, I think I could live with a `device(::AbstractGPUArray)` interface method.

Reply:
I've made it rely on array type instead of KA backend.